Implement cell reference handling

Fluglow opened the pull request:
* Implement tracking cell references in `referencesIn` and `referencesOut`
* When a cell's value is changed, its incoming references are used to update any other cells' values that have a reference to it
    * Since the core does no callbacks to the UI, this update is currently not reflected in the HTML
* Implement detecting circular cell references
* Clear cell value on invalid state

Yousif-FJ commented on tests/core/structure/sheet.test.ts:
```diff
@@ -20,4 +21,98 @@ describe("Sheet", () => {
   });
+
+  it("should create cells with references which are reflected by cell.referencesIn and cell.referencesOut", () => {
+    const setCell = function (col: number, row: number, value: string) {
+      const cInfo = sheet.setCellAt(col, row, value);
+      const cKey = sheet.columns
+        .get(cInfo.position.columnKey!)!
+        .cellIndex.get(cInfo.position.rowKey!)!;
+      return sheet.cellData.get(cKey)!;
+    };
```
Why is this function needed? It seems like a proxy function with a new return data but the returned data were never used

Fluglow replied:
The return data is used to populate the cells array below it, which is used to read the references of the Cell objects.

Yousif-FJ replied:
I would suggest this change, because it is not obvious what is going in here


```suggestion
    const getCellReference = (cInfo: CellInfo) => {
      const cKey = sheet.columns
        .get(cInfo.position.columnKey!)!
        .cellIndex.get(cInfo.position.rowKey!)!;
      return sheet.cellData.get(cKey)!;
    };

    const cells = [
      getCellReference(sheet.setCellAt(0, 0, "100")), // A1
      getCellReference(sheet.setCellAt(1, 0, "=A1")), // B1
      getCellReference(sheet.setCellAt(2, 2, "1")), // C3
      getCellReference(sheet.setCellAt(2, 0, "=A1 + B1 + C3")), // C1
    ];
```

Yousif-FJ commented on tests/core/structure/sheet.test.ts:
```diff
@@ -1,5 +1,6 @@
 import Cell from "../../../src/core/structure/cell/cell.ts";
+import { CellState } from "../../../src/core/structure/cell/cellState.ts";
 
 describe("Sheet", () => {
   let sheet: Sheet;
```
This test file needs to be broken down to different files. To me it seems like none of the issues are about `Sheet`

Yousif-FJ commented on tests/core/structure/sheet.test.ts:
```diff
@@ -20,4 +21,98 @@ describe("Sheet", () => {
+      [0, 2, "0"], // A3
+      [1, 2, "10"], // B3
+    ];
+    const cells = [];
```
What's the point of `cells` collection? It seens unused

Fluglow replied:
Oops, left over from a previous approach

Fluglow commented:
I've updated the code to use the event system to fully update incoming references. This required adding position information to reference tracking, but now chained references are also updated properly in the UI.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -62,25 +79,37 @@ export default class Sheet {
 
     if (cell) {
       cell.formula = value;
-      this.resolveCell(cell);
+      this.resolveCell(cell, colKey, rowKey);
     }
```
Some refactoring while reviewing the code

```suggestion
    if (cell) {
      if (value == "") {
        // Cell exists but is being cleared.
        this.deleteCell(colKey, rowKey);
      }
      cell.formula = value;
      this.resolveCell(cell, colKey, rowKey);
    } else {
      cell = this.createCell(colKey, rowKey, value);
    }
```

Fluglow replied:
This removes the `resolveCell` call for newly created cells which I think we should be doing. The default formatter could modify the value of the new cell even if its contents are empty.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -396,18 +439,107 @@ export default class Sheet {
+      return { valueChanged: false };
+
+    // Return the set of cells whose values has changed as a result of updating this cell.
+    const dirtyCells = new Map<CellKey, PositionInfo>();
```
We are not setting any value to this map. Is there something missing here?

Fluglow replied:
This was left over from a previous approach; `ResolveCellResult` is unused currently.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -377,7 +420,7 @@ export default class Sheet {
     const cell = new Cell();
     cell.formula = value;
```
It is a bit not intuitive that we set the formula.
Can we have a function for setting the value and encapsulate the cell values instead.

I wish that we also force the user to call `revolveCell` somehow.

Like maybe we can have a function `equeueValue`, which takes the new value, then a function called `fulfillValue` which take a `ResolveCellResult` as parameter and updates the references. We probably need another function for updating `referredCell` references, but I'm all with that. 

I just don't like how we currently crossing the abstraction line between cell and sheet. Almost like cell is not an object in the oop sense, but then you applied encapsulation in the `setState`. So I think it should be either a data record without encapsulation or make it into a real oop object and give it ownership of its state by only modifying it with functions.

Fluglow replied:
One approach we could take instead is move responsibility for the `Cell` fields to `ExpressionHandler` or some similar static, "all sheets"-scope structure. This way we could move the logic out from `Sheet` and maybe improve the clarity a bit. I don't think solving this issue is within the scope of this PR, since it's just adding reference tracking to the existing structures.

Yousif-FJ replied:
I don't think calling the variable `formula` is the right thing, I would prefer if revert it back to `value`, because it is a value, but it could be a formula, however it is not always a formula. I would be confused on how to set the value if the function is accepting a formula.

I'm fine with it as `value` if you don't want to make it into a function.

Fluglow replied:
No, the formula always corresponds to user input and the value is always the resolved value that's displayed in the cell in the UI. So semantically, the function accepting `value` doesn't make sense since we're always setting the formula.

Yousif-FJ replied:
We still have the value stored 2 times, in `value` and in `formula` do you think that this is fine?

Fluglow replied:
I think it's fine for now. I think it'd be premature optimization at this point to include a mechanism to use either the formula or value depending on if the formula is an expression.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -396,18 +435,114 @@ export default class Sheet {
+    if (!evalResult) {
+      const prevState = cell.state;
+      cell.setState(CellState.INVALID_EXPRESSION);
+      return prevState == CellState.OK; // Consider the cell's value changed if its state changes from OK to invalid.
     }
```
> Consider the cell's value changed if its state changes from OK to invalid.
I think this consideration doesn't hold true if we had a formula bar. For this statement, if I input an invalid statement then I edit the statement but it is also invalid it should still counts as a change, because we changed the formula, yeah the value hasn't changed but the formula did.

Fluglow replied:
Yes, but this is in `resolveCell`. We're only concerned with the value there, since that's what `resolveCell` is setting. You can see in `setCell` that an event is emitted regardless of the result. The return value is used when updating references, in which case it only matters if the cell's value changes, not the formula.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -396,18 +435,114 @@ export default class Sheet {
+
+    const valueChanged = cell.value != evalResult.value;
+    cell.value = evalResult.value;
+    this.updateCellReferences(cell, colKey, rowKey, valueChanged, evalResult);
```
This function internally resolve and emit functions, this name is under playing how many things this function is doing. I would suggest `UpdateAndEvaludateCellReferences`

Also, it is not ideal to have a function with that many parameters.

Fluglow replied:
> this name is under playing how many things this function is doing

I've renamed it to `handleCellReferenceChanges`; I think along with the comment this is descriptive enough.



> Also, it is not ideal to have a function with that many parameters.

Yes, but we need to pass a lot of context to the method. It's a result of splitting the method into two, which I think is still better.

Yousif-FJ commented on src/core/structure/sheet.ts:
```diff
@@ -396,18 +435,114 @@ export default class Sheet {
-    if (value == null) {
-      cell.state = CellState.INVALID_EXPRESSION;
-      return false;
+  /**
+   * Returns true if the value of the cell has changed.
+   */
+  private resolveCell(cell: Cell, colKey: ColumnKey, rowKey: RowKey): boolean {
```
This function is internally emitting events when handling other cells reference so I would say:
- Your comments is underplaying the significance of this function, I would prefer if it is removed. I'm not a fan of comments
- There is no need to return true/false when changes happen rather, I would suggest emitting the events directly from this function because you are already doing that inside of this function when updating other cells

Fluglow replied:
I would prefer to leave a comment describing the return value when its purpose isn't immediately obvious.
> There is no need to return true/false when changes happen rather, I would suggest emitting the events directly from 

I'll move the event emission logic into `resolveCell`; I agree it makes more sense there. The return value is still required though, since referring cells aren't affected if the cell's value doesn't change.
